lib/rules/indent.js

```diff
@@ -521,540 +393,486 @@ module.exports = {
+        * @param {ASTNode} node The node
+        * @returns {Token[]} A list of tokens and comments
+        */
+        function getTokensAndComments(node) {
+            if (!node) {
+                return sourceCode.tokensAndComments;
             }
 
-            if (node.body.length > 0) {
-                checkNodesIndent(node.body, indent);
+            if (!tokensByNodeCache.has(node)) {
+                tokensByNodeCache.set(node, getTokensAndComments(node.parent).filter(token => token.range[0] >= node.range[0] && token.range[1] <= node.range[1]));
             }
 
-            checkLastNodeLineIndent(node, indent - functionOffset);
+            return tokensByNodeCache.get(node);
         }
 
-
         /**
-         * Checks if the given node starts and ends on the same line
-         * @param {ASTNode} node The node to check
-         * @returns {boolean} Whether or not the block starts and ends on the same line.
+         * Check indentation for blocks
+         * @param {ASTNode} node node to check
+         * @returns {void}
          */
-        function isSingleLineNode(node) {
-            const lastToken = sourceCode.getLastToken(node),
-                startLine = node.loc.start.line,
-                endLine = lastToken.loc.end.line;
+        function addBlockIndent(node) {
 
-            return startLine === endLine;
-        }
+            let blockIndentLevel;
+            const tokens = getTokensAndComments(node);
 
-        /**
-         * Check to see if the first element inside an array is an object and on the same line as the node
-         * If the node is not an array then it will return false.
-         * @param {ASTNode} node node to check
-         * @returns {boolean} success/failure
-         */
-        function isFirstArrayElementOnSameLine(node) {
-            if (node.type === "ArrayExpression" && node.elements[0]) {
-                return node.elements[0].loc.start.line === node.loc.start.line && node.elements[0].type === "ObjectExpression";
+            if (node.parent && isOuterIIFE(node.parent)) {
+                blockIndentLevel = options.outerIIFEBody;
+            } else if (node.parent && (node.parent.type === "FunctionExpression" || node.parent.type === "ArrowFunctionExpression")) {
+                blockIndentLevel = options.FunctionExpression.body;
+            } else if (node.parent && node.parent.type === "FunctionDeclaration") {
+                blockIndentLevel = options.FunctionDeclaration.body;
             } else {
-                return false;
+                blockIndentLevel = 1;
             }
-        }
 
-        /**
-         * Check indent for array block content or object block content
-         * @param {ASTNode} node node to examine
-         * @returns {void}
-         */
-        function checkIndentInArrayOrObjectBlock(node) {
+            const tokenToMatchAgainst = getFirstTokenOfLine(tokens[0]) === tokens[0] ? tokens[0] : sourceCode.getFirstToken(node.parent);
```

Comment:
Wondering if this might deserve a comment (i.e., why not always use `sourceCode.getFirstToken(node.parent)`)?